src/common/src/util/stream_cancel.rs

```diff
@@ -0,0 +1,137 @@
+// Copyright 2023 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A library inspired by [stream-cancel](https://github.com/jonhoo/stream-cancel). This
+//! difference is that it allows you to send a value to original stream when cancelled, this is
+//! useful in several cases, for example when cancel query, it can be used to concat an cancel
+//! error to original stream.
+
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+use std::task::Poll;
+
+use futures::channel::oneshot::{channel, Receiver, Sender};
+use futures::stream::poll_fn;
+use futures::{Stream, StreamExt};
+
+pub struct Trigger {
+    sender: Sender<()>,
+}
+
+impl Trigger {
+    /// Used to trigger cancelling of stream.
+    pub fn abort(self) {
+        if self.sender.send(()).is_err() {
+            tracing::debug!("Trigger receiver closed, ignoring...");
+        }
+    }
+}
+
+pub struct Tripwire<R> {
+    receiver: Receiver<()>,
+    cancel_result: Box<dyn FnOnce() -> R + Send>,
+}
+
+pub fn stream_tripwire<F, R>(cancel_result: F) -> (Trigger, Tripwire<R>)
+where
+    F: FnOnce() -> R + 'static + Send,
+{
+    let (sender, receiver) = channel();
+    (
+        Trigger { sender },
+        Tripwire {
+            receiver,
+            cancel_result: Box::new(cancel_result),
+        },
+    )
+}
+
+pub fn cancellable_stream<S>(
```

Comment:
There's an `Abortable` in `futures`. Does it meet our requirements?

Reply:
No, I want to user to receive query cancel error when query cancelled.

Reply:
`Abortable` will wrap the original stream and yields `Err(Aborted)` after cancelling.

https://docs.rs/futures/latest/futures/stream/struct.Abortable.html#impl-Abortable%3CT%3E

Reply:
It just stop stream

Reply:
Don't quite get that. Doesn't `take_until` also stop the stream? If you mean doing some "post work" after it's aborted, simply map the `Err` then we can get it done there.

Reply:
I mean when we abort stream, no error returned

Reply:
It's just the interface design used as a signal of abortion of `futures::Abortable`. 😂 If we don't want the `Err` itself, transform or discard it before turning the stream into an opaque type.

Reply:
I think it's hard to design the api.